EntityId cache lookup for evmAddress/Alias can throw Recursive Update Exception

### Description

The evm address and alias cacheLookup method in EntityIdServiceImpl can result in a recursive update exception.

Here is the exception from testnet: 

```
2024-08-15 13:12:22.988	2024-08-15T18:12:22.988Z ERROR pool-11-thread-2 c.h.m.i.util.Utility Recoverable error. Invalid itemizedTransfer entity id at 1723745537096530003 
2024-08-15 13:12:22.988	2024-08-15T18:12:22.984Z ERROR pool-11-thread-2 c.h.m.i.util.Utility Recoverable error. Error looking up alias or EVM address <ByteString@54cf70ae size=35 contents=":!\003\223\026\203\210G\273W\t\027\276Jl\001\373\362O\214\301$\211(\331\a/\257\362\345\272\n\315\363\200"> from cache org.springframework.cache.Cache$ValueRetrievalException: Value for key '<ByteString@54cf70ae size=35 contents=":!\003\223\026\203\210G\273W\t\027\276Jl\001\373\362O\214\301$\211(\331\a/\257\362\345\272\n\315\363\200">' could not be loaded using 'com.hedera.mirror.importer.domain.EntityIdServiceImpl$$Lambda/0x00007be33d42fc18@4e88d541'
	at org.springframework.cache.caffeine.CaffeineCache$LoadFunction.apply(CaffeineCache.java:240)
	at com.github.benmanes.caffeine.cache.LocalCache.lambda$statsAware$2(LocalCache.java:167)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.lambda$doComputeIfAbsent$14(BoundedLocalCache.java:2688)
	at java.base/java.util.concurrent.ConcurrentHashMap.compute(Unknown Source)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2686)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2669)
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:112)
	at com.github.benmanes.caffeine.cache.LocalManualCache.get(LocalManualCache.java:62)
	at org.springframework.cache.caffeine.CaffeineCache.get(CaffeineCache.java:135)
	at com.hedera.mirror.importer.domain.EntityIdServiceImpl.cacheLookup(EntityIdServiceImpl.java:121)
	at com.hedera.mirror.importer.domain.EntityIdServiceImpl.lookup(EntityIdServiceImpl.java:67)
	at com.hedera.mirror.importer.parser.record.entity.EntityRecordItemListener.processItemizedTransfers(EntityRecordItemListener.java:240)
	at com.hedera.mirror.importer.parser.record.entity.EntityRecordItemListener.onItem(EntityRecordItemListener.java:150)
	at com.hedera.mirror.importer.parser.record.entity.EntityRecordItemListener.onItem(EntityRecordItemListener.java:77)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.lambda$doParse$0(RecordFileParser.java:174)
	at java.base/java.util.ArrayList.forEach(Unknown Source)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:166)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:52)
	at com.hedera.mirror.importer.parser.AbstractStreamFileParser.parse(AbstractStreamFileParser.java:104)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.parse(RecordFileParser.java:127)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:354)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:379)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:92)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:344)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:233)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:113)
	at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:162)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:720)
	at com.hedera.mirror.importer.parser.record.RecordFileParser$$SpringCGLIB$$0.parse(<generated>)
	at com.hedera.mirror.importer.downloader.BatchStreamFileNotifier$StreamFileSubscriber.handle(BatchStreamFileNotifier.java:134)
	at com.hedera.mirror.importer.downloader.BatchStreamFileNotifier$StreamFileSubscriber.run(BatchStreamFileNotifier.java:107)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.IllegalStateException: Recursive update
	at java.base/java.util.concurrent.ConcurrentHashMap.transfer(Unknown Source)
	at java.base/java.util.concurrent.ConcurrentHashMap.addCount(Unknown Source)
	at java.base/java.util.concurrent.ConcurrentHashMap.compute(Unknown Source)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2686)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2669)
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:112)
	at com.github.benmanes.caffeine.cache.LocalManualCache.get(LocalManualCache.java:62)
	at org.springframework.cache.caffeine.CaffeineCache.get(CaffeineCache.java:135)
	at com.hedera.mirror.importer.domain.EntityIdServiceImpl.findByPublicKeyAlias(EntityIdServiceImpl.java:207)
	at com.hedera.mirror.importer.domain.EntityIdServiceImpl.lambda$lookup$0(EntityIdServiceImpl.java:71)
	at org.springframework.cache.caffeine.CaffeineCache$LoadFunction.apply(CaffeineCache.java:237)
	... 43 common frames omitted
```

### Steps to reproduce

1. Ingest the record stream from that consensus timestamp from testnet.

### Additional context

_No response_

### Hedera network

testnet

### Version

v.0.111.0

### Operating system

None

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/EntityIdServiceImpl.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/EntityIdServiceImpl.java
@@ -16,6 +16,7 @@
 
 package com.hedera.mirror.importer.domain;
 
+import static com.hedera.mirror.common.util.DomainUtils.EVM_ADDRESS_LENGTH;
 import static com.hedera.mirror.common.util.DomainUtils.fromBytes;
 import static com.hedera.mirror.common.util.DomainUtils.toBytes;
 import static com.hedera.mirror.importer.config.CacheConfiguration.CACHE_ALIAS;
@@ -64,13 +65,24 @@ public class EntityIdServiceImpl implements EntityIdService {
 
         return switch (accountId.getAccountCase()) {
             case ACCOUNTNUM -> Optional.ofNullable(EntityId.of(accountId));
-            case ALIAS -> cacheLookup(accountId.getAlias(), () -> {
+            case ALIAS -> {
+                long shard = accountId.getShardNum();
+                long realm = accountId.getRealmNum();
                 byte[] alias = toBytes(accountId.getAlias());
-                return alias.length == DomainUtils.EVM_ADDRESS_LENGTH
-                        ? findByEvmAddress(alias, accountId.getShardNum(), accountId.getRealmNum())
-                        : findByPublicKeyAlias(alias, accountId.getShardNum(), accountId.getRealmNum());
-            });
 
+                yield cacheLookup(
+                                accountId.getAlias(),
+                                () -> alias.length == EVM_ADDRESS_LENGTH
+                                        ? findByEvmAddress(alias, shard, realm)
+                                        : entityRepository.findByAlias(alias).map(EntityId::of))
+                        .or(() -> {
+                            if (alias.length == EVM_ADDRESS_LENGTH) {
+                                return Optional.empty();
+                            }
+                            var evmAddress = aliasToEvmAddress(alias);
+                            return cacheLookup(fromBytes(evmAddress), () -> findByEvmAddress(evmAddress, shard, realm));
+                        });
+            }
             default -> {
                 Utility.handleRecoverableError(
                         "Invalid Account Case for AccountID {}: {}", accountId, accountId.getAccountCase());
@@ -118,6 +130,10 @@ public class EntityIdServiceImpl implements EntityIdService {
 
     private Optional<EntityId> cacheLookup(ByteString key, Callable<Optional<EntityId>> loader) {
         try {
+            if (key == null || key.equals(ByteString.EMPTY)) {
+                return Optional.empty();
+            }
+
             return cache.get(key, loader);
         } catch (Cache.ValueRetrievalException e) {
             Utility.handleRecoverableError("Error looking up alias or EVM address {} from cache", key, e);
@@ -181,29 +197,5 @@ public class EntityIdServiceImpl implements EntityIdService {
         }
 
         return id;
-    }
-
-    private Optional<EntityId> findByPublicKeyAlias(byte[] alias, long shardNum, long realmNum) {
-        var encodedId = entityRepository.findByAlias(alias);
-        if (encodedId.isPresent()) {
-            return Optional.ofNullable(EntityId.of(encodedId.get()));
-        }
-
-        // Try to fall back to the 20-byte evm address recovered from the ECDSA secp256k1 alias
-        var evmAddress = aliasToEvmAddress(alias);
-        if (evmAddress == null) {
-            Utility.handleRecoverableError("Unable to find entity for alias {}", Hex.encodeHexString(alias));
-            return Optional.empty();
-        }
-
-        if (log.isDebugEnabled()) {
-            log.debug(
-                    "Trying to find entity by evm address {} recovered from public key alias {}",
-                    Hex.encodeHexString(evmAddress),
-                    Hex.encodeHexString(alias));
-        }
-
-        // Check cache first in case the 20-byte evm address hasn't persisted to db
-        return cache.get(fromBytes(evmAddress), () -> findByEvmAddress(evmAddress, shardNum, realmNum));
     }
 }

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/EntityIdServiceImpl.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/EntityIdServiceImpl.java
@@ -71,12 +71,9 @@ public class EntityIdServiceImpl implements EntityIdService {
                 long shard = accountId.getShardNum();
                 long realm = accountId.getRealmNum();
                 byte[] alias = toBytes(accountId.getAlias());
-
                 yield alias.length == EVM_ADDRESS_LENGTH
                         ? cacheLookup(accountId.getAlias(), () -> findByEvmAddress(alias, shard, realm))
-                        : cacheLookup(accountId.getAlias(), () -> entityRepository
-                                        .findByAlias(alias)
-                                        .map(EntityId::of))
+                        : cacheLookup(accountId.getAlias(), () -> findByAlias(alias))
                                 .or(() -> findByAliasEvmAddress(alias, shard, realm));
             }
             default -> {
@@ -126,10 +123,6 @@ public class EntityIdServiceImpl implements EntityIdService {
 
     private @Nonnull Optional<EntityId> cacheLookup(ByteString key, Callable<Optional<EntityId>> loader) {
         try {
-            if (key == null || key.equals(ByteString.EMPTY)) {
-                return Optional.empty();
-            }
-
             return Objects.requireNonNullElse(cache.get(key, loader), Optional.empty());
         } catch (Cache.ValueRetrievalException e) {
             Utility.handleRecoverableError("Error looking up alias or EVM address {} from cache", key, e);
@@ -195,8 +188,26 @@ public class EntityIdServiceImpl implements EntityIdService {
         return id;
     }
 
-    private Optional<EntityId> findByAliasEvmAddress(byte[] alias, long shard, long realm) {
+    private Optional<EntityId> findByAlias(byte[] alias) {
+        return entityRepository.findByAlias(alias).map(EntityId::of);
+    }
+
+    // Try to fall back to the 20-byte evm address recovered from the ECDSA secp256k1 alias
+    private Optional<EntityId> findByAliasEvmAddress(byte[] alias, long shardNum, long realmNum) {
         var evmAddress = aliasToEvmAddress(alias);
-        return cacheLookup(fromBytes(evmAddress), () -> findByEvmAddress(evmAddress, shard, realm));
+        if (evmAddress == null) {
+            Utility.handleRecoverableError("Unable to find entity for alias {}", Hex.encodeHexString(alias));
+            return Optional.empty();
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug(
+                    "Trying to find entity by evm address {} recovered from public key alias {}",
+                    Hex.encodeHexString(evmAddress),
+                    Hex.encodeHexString(alias));
+        }
+
+        // Check cache first in case the 20-byte evm address hasn't persisted to db
+        return cacheLookup(fromBytes(evmAddress), () -> findByEvmAddress(evmAddress, shardNum, realmNum));
     }
 }

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/EntityIdServiceImpl.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/EntityIdServiceImpl.java
@@ -33,7 +33,9 @@ import com.hedera.mirror.importer.repository.EntityRepository;
 import com.hedera.mirror.importer.util.Utility;
 import com.hederahashgraph.api.proto.java.AccountID;
 import com.hederahashgraph.api.proto.java.ContractID;
+import jakarta.annotation.Nonnull;
 import jakarta.inject.Named;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.Callable;
 import java.util.function.Function;
@@ -70,18 +72,12 @@ public class EntityIdServiceImpl implements EntityIdService {
                 long realm = accountId.getRealmNum();
                 byte[] alias = toBytes(accountId.getAlias());
 
-                yield cacheLookup(
-                                accountId.getAlias(),
-                                () -> alias.length == EVM_ADDRESS_LENGTH
-                                        ? findByEvmAddress(alias, shard, realm)
-                                        : entityRepository.findByAlias(alias).map(EntityId::of))
-                        .or(() -> {
-                            if (alias.length == EVM_ADDRESS_LENGTH) {
-                                return Optional.empty();
-                            }
-                            var evmAddress = aliasToEvmAddress(alias);
-                            return cacheLookup(fromBytes(evmAddress), () -> findByEvmAddress(evmAddress, shard, realm));
-                        });
+                yield alias.length == EVM_ADDRESS_LENGTH
+                        ? cacheLookup(accountId.getAlias(), () -> findByEvmAddress(alias, shard, realm))
+                        : cacheLookup(accountId.getAlias(), () -> entityRepository
+                                        .findByAlias(alias)
+                                        .map(EntityId::of))
+                                .or(() -> findByAliasEvmAddress(alias, shard, realm));
             }
             default -> {
                 Utility.handleRecoverableError(
@@ -128,13 +124,13 @@ public class EntityIdServiceImpl implements EntityIdService {
         return doLookups(contractIds, this::lookup);
     }
 
-    private Optional<EntityId> cacheLookup(ByteString key, Callable<Optional<EntityId>> loader) {
+    private @Nonnull Optional<EntityId> cacheLookup(ByteString key, Callable<Optional<EntityId>> loader) {
         try {
             if (key == null || key.equals(ByteString.EMPTY)) {
                 return Optional.empty();
             }
 
-            return cache.get(key, loader);
+            return Objects.requireNonNullElse(cache.get(key, loader), Optional.empty());
         } catch (Cache.ValueRetrievalException e) {
             Utility.handleRecoverableError("Error looking up alias or EVM address {} from cache", key, e);
             return Optional.empty();
@@ -197,5 +193,10 @@ public class EntityIdServiceImpl implements EntityIdService {
         }
 
         return id;
+    }
+
+    private Optional<EntityId> findByAliasEvmAddress(byte[] alias, long shard, long realm) {
+        var evmAddress = aliasToEvmAddress(alias);
+        return cacheLookup(fromBytes(evmAddress), () -> findByEvmAddress(evmAddress, shard, realm));
     }
 }

--- a/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/domain/EntityIdServiceImplTest.java
+++ b/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/domain/EntityIdServiceImplTest.java
@@ -385,11 +385,19 @@ class EntityIdServiceImplTest extends ImporterIntegrationTest {
     }
 
     @Test
-    void lookupAccountWithEvmAddress() {
+    void lookupAccountEvmAddress() {
         AccountID accountId = AccountID.newBuilder()
                 .setAlias(DomainUtils.fromBytes(PARSABLE_EVM_ADDRESS))
                 .build();
         assertThat(entityIdService.lookup(accountId)).hasValue(EntityId.of(100));
+    }
+
+    @Test
+    void lookupAccountEvmAddressNotFound() {
+        AccountID accountId = AccountID.newBuilder()
+                .setAlias(DomainUtils.fromBytes(domainBuilder.evmAddress()))
+                .build();
+        assertThat(entityIdService.lookup(accountId)).isNotPresent();
     }
 
     private AccountID getProtoAccountId(Entity account) {

